plugins/redfish/fu-ipmi-device.c

```diff
@@ -0,0 +1,573 @@
+					req,
+					sizeof(req),
+					NULL, /* resp */
+					0,
+					NULL,
+					FU_IPMI_DEVICE_TIMEOUT,
+					error)) {
+		g_prefix_error(error, "failed to set user %02x enable: ", user_id);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_ipmi_device_set_user_password(FuIpmiDevice *self,
+				 guint8 user_id,
+				 const gchar *password,
+				 GError **error)
+{
+	guint8 req[0x12] = {user_id, IPMI_PASSWORD_SET_PASSWORD};
+	gsize password_sz;
+
+	g_return_val_if_fail(FU_IS_IPMI_DEVICE(self), FALSE);
+	g_return_val_if_fail(user_id != 0x0, FALSE);
+	g_return_val_if_fail(password != NULL, FALSE);
+
+	/* copy into buffer */
+	password_sz = strlen(password);
+	if (!fu_memcpy_safe(req,
+			    sizeof(req),
+			    0x2, /* dst */
+			    (guint8 *)password,
+			    password_sz,
+			    0x0, /* src */
+			    password_sz,
+			    error)) {
+		g_prefix_error(error, "password invalid: ");
+		return FALSE;
+	}
+
+	/* run transaction */
+	if (!fu_ipmi_device_transaction(self,
+					IPMI_NETFN_APP_REQUEST,
+					IPMI_SET_USER_PASSWORD,
+					req,
+					sizeof(req),
+					NULL, /* resp */
+					0,
+					NULL,
+					FU_IPMI_DEVICE_TIMEOUT,
+					error)) {
+		g_prefix_error(error, "failed to set user %02x password: ", user_id);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_ipmi_device_set_user_priv(FuIpmiDevice *self,
+			     guint8 user_id,
+			     guint8 priv_limit,
+			     guint8 channel,
+			     GError **error)
+{
+	const guint8 req[] = {channel, user_id, priv_limit, 0x0};
+
+	g_return_val_if_fail(FU_IS_IPMI_DEVICE(self), FALSE);
+	g_return_val_if_fail(user_id != 0x0, FALSE);
+	g_return_val_if_fail(channel <= 0x0F, FALSE);
+	g_return_val_if_fail(priv_limit <= 0x0F, FALSE);
+
+	/* run transaction */
+	if (!fu_ipmi_device_transaction(self,
+					IPMI_NETFN_APP_REQUEST,
+					IPMI_SET_USER_ACCESS,
+					req,
+					sizeof(req),
+					NULL, /* resp */
+					0,
+					NULL,
+					FU_IPMI_DEVICE_TIMEOUT,
+					error)) {
+		g_prefix_error(error,
+			       "failed to set user %02x privs of 0x%02x, 0x%02x: ",
+			       user_id,
+			       priv_limit,
+			       channel);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_ipmi_device_init(FuIpmiDevice *self)
+{
+	fu_device_set_physical_id(FU_DEVICE(self), "/dev/ipmi0");
+	fu_device_set_name(FU_DEVICE(self), "IPMI");
+	fu_device_set_summary(FU_DEVICE(self), "Intelligent Platform Management Interface");
+	fu_device_add_icon(FU_DEVICE(self), "computer");
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_INTERNAL);
+}
+
+static void
+fu_ipmi_device_class_init(FuIpmiDeviceClass *klass)
+{
+	FuDeviceClass *klass_device = FU_DEVICE_CLASS(klass);
+	klass_device->setup = fu_ipmi_device_setup;
+	klass_device->to_string = fu_ipmi_device_to_string;
+}
+
+FuIpmiDevice *
+fu_ipmi_device_new(FuContext *ctx)
+{
+	FuIpmiDevice *self;
+	self = g_object_new(FU_TYPE_IPMI_DEVICE, "context", ctx, "device-file", "/dev/ipmi0", NULL);
```

Comment:
can't this also be `/dev/ipmidev/0`?  Or are those just symlinks?

Reply:
I don't see any of `/dev/ipmidev` at all on RHEL/Fedora.

Reply:
I don't know the context of it - but I see a few different things tried in ipmitool...
https://github.com/ipmitool/ipmitool/blob/d674bfdc9a2447706ea9b0b943c98bcb0bbe80d2/src/plugins/open/open.c#L102

Reply:
I had no idea they existed. I don't think I want to support more than one BMC device (!!!) but I guess it makes sense to search for /dev/ipmi/0 and /dev/ipmidev/0 too

Reply:
Fixed.